allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,274 @@
+            output_dict["loss"] = loss
+            # TODO: Define metrics
+        return output_dict
+
+    def _get_encoder_outputs(self, source_tokens: Dict[str, torch.LongTensor]) -> torch.LongTensor:
+        embedded_input = self._source_embedder(source_tokens)
+        source_mask = get_text_field_mask(source_tokens)
+        encoded_text = self._encoder(embedded_input, source_mask)
+        return encoded_text
+
+    def _prepare_decode_step_input(self, input_indices: torch.LongTensor,
+                                   last_decoder_output: torch.LongTensor = None,
+                                   encoder_outputs: torch.LongTensor = None) -> torch.LongTensor:
+        """
+        Given the input indices for the current timestep of the decoder, and all the encoder outputs,
+        compute the input at the current timestep.
+        Note: This method is agnostic to whether the indices are gold indices or the predictions made by the
+        decoder at the last timestep. So, this can be used even if we're doing some kind of scheduled sampling.
+
+        If we're not using attention, the output of this method is just an embedding of the input indices.
+        If we are, the output will be a concatentation of the embedding and an attended average of the encoder
+        inputs.
+
+        Parameters
+        ----------
+        input_indices : torch.LongTensor
+            Indices of either the gold inputs to the decoder or the predicted labels from the previous timestep.
+        last_decoder_output : torch.LongTensor, optional
+            Output of from the decoder at the last time step. Needed only if using attention.
+        encoder_outputs : torch.LongTensor
+            Encoder outputs from all time steps. Needed only if using attention.
+        """
+        # input_indices : (batch_size,)  since we are processing these one timestep at a time.
+        # (batch_size, enc_embedding_dim)
+        embedded_input = self._target_embedder(input_indices)
+        if self._use_attention:
+            if last_decoder_output is None or encoder_outputs is None:
+                raise ConfigurationError("Last decoder output and all encoder outputs are needed to compute "
+                                         "attention weights.")
+            # encoder_outputs : (batch_size, input_sequence_length, enc_output_dim)
+            # (batch_size, input_sequence_length)
+            input_weights = self._decoder_attention(last_decoder_output, encoder_outputs)
+            # (batch_size, enc_output_dim)
+            attended_input = weighted_sum(encoder_outputs, input_weights)
+            # (batch_size, enc_output_dim + enc_embedding_dim)
+            return torch.cat((attended_input, embedded_input), -1)
+        else:
+            return embedded_input
+
+    def _predict_step(self, # pylint: disable=no-self-use
+                      target_scores: torch.LongTensor) -> Dict[str, torch.LongTensor]:
```

Comment:
Not sure the dictionary return type is buying you anything here over a simple tuple.